### PR TITLE
GODRIVER-2366 Restrict minPoolSize to not exceeding maxPoolSize

### PR DESCRIPTION
**Pull request by qingyang-hu**
[GODRIVER-2366](https://jira.mongodb.org/browse/GODRIVER-2366)

I raise a panic from the initializer of `serverConfig` when `minPoolSize` > `maxPoolSize` because it is an unsupported behavior rather than an error that needs to be handled in my opinion. Please let me know if it is a proper solution.

Also, I added some test cases to cover the changes.

### CODE REVIEW

File: x/mongo/driver/topology/pool.go
Comment (benjirewis) [inline x/mongo/driver/topology/pool.go:179]: Can you leave a comment above this line explaining that `minSize` should never exceed `maxSize`?

File: x/mongo/driver/topology/pool_test.go
Comment (benjirewis) [inline x/mongo/driver/topology/pool_test.go:27]: Do subtests of tests marked with `t.Parallel()` also need `t.Parallel()`?

File: x/mongo/driver/topology/server_options.go
Comment (benjirewis) [inline x/mongo/driver/topology/server_options.go:64]: Hmm I don't love the `panic` here. I think generally, we want to return an informative error to the user as soon as possible as opposed to `panic`ing deeper within the `x` package. I'm wondering if we can set the error on the returned `ClientOptions` object in [`ApplyURI`](https://github.com/mongodb/mongo-go-driver/blob/master/mongo/options/clientoptions.go#L236) instead. Something like:

```
// Verify that if both minPoolSize and maxPoolSize are set, minPoolSize < maxPoolSize.
if cs.MinPoolSizeSet && cs.MaxPoolSizeSet && cs.MinPoolSize >= cs.MaxPoolSize {
  c.err = fmt.Errorf("set minPoolSize %d exceeds set maxPoolSize %d", cs.MinPoolSize, cs.MaxPoolSize)
  return c
}
```

File: x/mongo/driver/topology/server_options_test.go
Comment (benjirewis) [inline x/mongo/driver/topology/server_options_test.go:26]: May have to refactor these tests if we remove the `panic` and error within `ApplyURI`.

**Review by benjirewis — COMMENTED** [review]
Looking great! I'm a little concerned about the `panic` ⚠️ .

File: x/mongo/driver/topology/pool_test.go
Comment (qingyang-hu) [reply x/mongo/driver/topology/pool_test.go:27]: We need `Parallel()` in every subtest, basically meaning `Parallel()` is paired with `Run()` to run the tests in parallel, in my understanding. However, I agree the `Parallel()` calls are not necessary for the subtests under "newPool" considering they are lightweight.

Comment (benjirewis) [reply x/mongo/driver/topology/pool_test.go:27]: Sounds good. Whether or not the `newPool` tests _need_ to be `Parallel()`, I don't have a strong opinion. Unless I'm mistaken, I don't think our CI yet runs tests in parallel. Parallelizing the test suite may be done with GODRIVER-2093.

Comment (benjirewis) [reply x/mongo/driver/topology/pool_test.go:27]: Awesome; thanks so much for that explanation.

File: x/mongo/driver/topology/server_options.go
Comment (matthewdale) [inline x/mongo/driver/topology/server_options.go:70]: The validation error message should reference the user-settable config value name. For example:
> minPoolSize must be less than or equal to maxPoolSize, got minPoolSize=%d maxPoolSize=%d

**Review by matthewdale — COMMENTED** [review]
Nice! One suggestion to improve the validation error message.

**Comment by qingyang-hu** [comment]
I add a validation function in [server_options.go](https://github.com/qingyang-hu/mongo-go-driver/blob/478e9a63aeecde62480dddc88d48a467f9ed6082/x/mongo/driver/topology/server_options.go#L45) because we set the default `maxConns` value [here](https://github.com/qingyang-hu/mongo-go-driver/blob/478e9a63aeecde62480dddc88d48a467f9ed6082/x/mongo/driver/topology/server_options.go#L53). It is difficult to validate the `minConns` in the `ClientOptions` when it exceeds the default `maxConns` value. Meanwhile, I call the validation function inside `Client` [configure()](https://github.com/qingyang-hu/mongo-go-driver/blob/478e9a63aeecde62480dddc88d48a467f9ed6082/mongo/client.go#L684), which is called by the `Client` [initializer](https://github.com/qingyang-hu/mongo-go-driver/blob/478e9a63aeecde62480dddc88d48a467f9ed6082/mongo/client.go#L132), where a proper error can be returned to the caller after the `ClientOptions` being passed in.

File: x/mongo/driver/topology/server_options.go
Comment (qingyang-hu) [reply x/mongo/driver/topology/server_options.go:70]: It's what I hesitate in. The setting function is named by `WithMinConnections` and `WithMaxConnecting`.

Comment (matthewdale) [reply x/mongo/driver/topology/server_options.go:70]: Yeah, it is weird to reference a different field name than the one you're actually validating. Per our previous conversation, it might be less weird logic to move the default `maxConns` out of `newServerConfig` and into the client configuration options. That way you can do the validation and error messaging from the `ClientOptions` validator rather than a new server config validator.

**Comment by qingyang-hu** [comment]
I will merge changes for ClientOptions.Validate(), @matthewdale

File: mongo/client.go
Comment (matthewdale) [inline mongo/client.go:358]: Optional: Consider moving this `MaxPoolSize` default handling block down to where the existing `if opts.MaxPoolSize != nil` block is to keep all the `MaxPoolSize` option logic in the same place.

**Review by matthewdale — APPROVED** [review]
Looks good 👍 

P.S. I'm going to hold off merging https://github.com/mongodb/mongo-go-driver/pull/938 until this is merged to prevent merge conflicts in this PR because the merge conflicts will be easier to resolve in that PR.

File: mongo/client_test.go
Comment (matthewdale) [inline mongo/client_test.go:188]: Optional: Consider using keyed fields and breaking out the test case structs onto multiple lines to make them easier to read.

E.g.
```go
{
	name: "minPoolSize < default maxPoolSize",
	opts: options.Client().ApplyURI("mongodb://localhost:27017/?minPoolSize=64"),
	err: nil,
},
```

File: mongo/client.go
Comment (qingyang-hu) [reply mongo/client.go:358]: Default options (`maxPoolSize` currently) need to be set first so they can all be checked in `Validate()`, and `Validate()` needs to be called before actual value modifications underneath. Similarly, I used `SetMaxPoolSize()` to indicate it an external setting somehow.

File: mongo/client.go
Comment (benjirewis) [inline mongo/client.go:354]: Are we expecting more default options in the future?
```suggestion
	var defaultOptions int
```
Firstly, I think you can just use the zero-value to initialize here. But, are we expecting more default options in the future? Even if we were, I might have a hard time remembering to increment this value for new ones. I think you can just require `len(serverOpts) > 3` on L700 and note that `// server options: WithClock, WithConnectionOptions and WithMaxPoolSize` are included in the default server options.

File: mongo/client.go
Comment (benjirewis) [reply mongo/client.go:359]: Setting here seems fine to me in order to call `Validate` after and error quickly.

File: mongo/client_test.go
Comment (benjirewis) [inline mongo/client_test.go:182]: Nice tests! I think it's good to test the same functionality through `ApplyURI` and the `Set*` methods.

File: x/mongo/driver/topology/pool_test.go
Comment (benjirewis) [inline x/mongo/driver/topology/pool_test.go:34]: I think you want a different name for this test. Something like `minPoolSize can exceed maxPoolSize of 0`? The assertion error message might have to change, too.

File: mongo/client.go
Comment (qingyang-hu) [reply mongo/client.go:354]: The original logic on [L700](https://github.com/mongodb/mongo-go-driver/blob/5c6942e0f0ee496dc0fcbca2d5a5cb597ada3154/mongo/client.go#L700) confuses me. It seems not even accept a customer `maxPoolSize` setting (excluding the default option logic) to me. If this is the case, a counter is more clear than twisted `if` conditions in my opinion even if it is the only default value.

**Review by benjirewis — APPROVED** [review]
LGTM! I think the final commit message should probably be `GODRIVER-2366 Restrict minPoolSize to not exceed maxPoolSize.`.

**Comment by qingyang-hu** [comment]
Thank you, all. @matthewdale, you can merge [#938](https://github.com/mongodb/mongo-go-driver/pull/938)

**Pull request merged**
